tket/src/WeightSubgrMono/include/WeightSubgrMono/Searching/SearchBranch.hpp

```diff
@@ -0,0 +1,147 @@
+   * the very first search or not.
+   */
+  bool move_down_has_been_called() const;
+
+  /** Move up one level, but do NOT reduce the node.
+   * Returns false if finished, i.e. the search is OVER!
+   */
+  bool backtrack();
+
+  /** ASSUMING that the current node has been fully reduced,
+   * make the given assignment [which must be valid; hence this cannot fail]
+   * and move down one level to a new node.
+   * The new node, however, will not yet be reduced.
+   * @param p_vertex A pattern vertex, pv.
+   * @param t_vertex A target vertex, tv. We are making the new assignment
+   * pv->tv.
+   */
+  void move_down(VertexWSM p_vertex, VertexWSM t_vertex);
+
+  /** Reduce the current node as much as possible,
+   * until it is valid (or invalid), i.e. not in a "partial" state;
+   * if successfully reduced, we are confident that
+   * there are no edge/vertex clashes, the current weight
+   * is up to date,...
+   * Does NOT take any further action with full/partial solutions, though.
+   * @param shared_data The data shared across ALL search branches.
+   * @param max_weight The maximum permitted weight (scalar product) for any
+   * solution.
+   * @return What happened with the reduction.
+   */
+  ReductionResult reduce_current_node(
+      SharedData& shared_data, WeightWSM max_weight);
+
+  /** Simply returns the read-only accessor for the current search node. */
+  const SearchNodeWrapper& get_current_node_wrapper() const;
+
+  typedef std::vector<EnrichedNode> EnrichedNodes;
+
+  /** Returns all the data for the search so far, and the current level
+   * by reference (i.e., index of current enriched node;
+   * NOT the number of nodes, which is level+1).
+   * The CALLER can decide what to do with partial/complete solutions, etc.
+   * @param level The current level, returned by reference.
+   * @return The current list of all search nodes. NOTE that there may be extra
+   * unused data on the end, which should be ignored; it would be wasteful to
+   * keep resizing and deallocating/reallocating memory for search nodes. This
+   * is why the current level is also returned.
+   */
+  const EnrichedNodes& get_data(std::size_t& level) const;
+
+  /** Returns the current assignments pv->tv in the search.
+   * @return All current pv->tv assignments made in this search branch down to
+   * and including the current node.
+   */
+  const Assignments& get_assignments() const;
+
+  /** Should be rare; returns a writeable version of the assignments.
+   * @return All current pv->tv assignments made in this search branch,
+   * non-const.
+   */
+  Assignments& get_assignments_mutable();
+
+  /** If we've decided that PV->TV is an impossible assignment,
+   * try to erase TV from every Dom(PV). However, for simplicity,
+   * do not introduce extra assignments, i.e. if Dom(PV) = { TV, u }
+   * at some level, leave Dom(PV) unchanged, as otherwise we'd need
+   * more complicated code to force PV=u; it will naturally take care
```

Comment:
As the search progresses, the next time PV -> TV is attempted, it will be immediately rejected, and then backtracking will erase that possibility from the data.

So, the extra algorithmic complexity of taking care of it earlier is not worth it, I think. (But I'm not going to spend time implementing it to see!)

Reply:
Well, actually, thinking about it more, maybe it is worth it; but it's complicated, I'm leaving it for now.

Everything is a balancing act; the crucial thing is correctness. Sometimes you add extra computations to speed it up, but they take longer than the time saved!

Reply:
Thannks for the explanation